consensus/state_processing/src/common/initialize_progressive_balances_cache.rs

```diff
@@ -0,0 +1,61 @@
+use crate::per_epoch_processing::altair::ParticipationCache;
+use std::borrow::Cow;
+use types::{BeaconState, BeaconStateError, ChainSpec, EthSpec};
+
+/// Initializes the `ProgressiveBalancesCache` cache using balance values from the
+/// `ParticipationCache`. If the optional `&ParticipationCache` is not supplied, it will be computed
+/// from the `BeaconState`.
+pub fn initialize_progressive_balances_cache<E: EthSpec>(
+    state: &mut BeaconState<E>,
+    maybe_participation_cache: Option<&ParticipationCache>,
+    spec: &ChainSpec,
+) -> Result<(), BeaconStateError> {
+    if !is_progressive_balances_enabled(state)
+        || state.progressive_balances_cache().is_initialized()
+    {
+        return Ok(());
+    }
+
+    let participation_cache = match maybe_participation_cache {
+        Some(cache) => Cow::Borrowed(cache),
+        None => Cow::Owned(ParticipationCache::new(state, spec)?),
+    };
+
+    // FIXME[JC]: Converts value to 0 if it is the same as `EFFECTIVE_BALANCE_INCREMENT`.
+    // `ParticipationCache` methods return `EFFECTIVE_BALANCE_INCREMENT` (1,000,000,000)
+    // when the balance is 0, and this breaks our calculation.
+    let handle_zero_effective_balance = |val| {
+        if val == spec.effective_balance_increment {
+            0
+        } else {
+            val
+        }
+    };
```

Comment:
It might be neater if we exposed methods on `ParticipationCache` like `previous_epoch_target_attesting_balance_raw()` and `current_epoch_target_attesting_balance_raw()`. Then the cache could be initialised with the correct balances without any reverse engineering

I think we should also make it so that the progressive balances cache returns `max(balance, EFFECTIVE_BALANCE_INCREMENT)` just like the participation cache. Both are meant to implement the semantics of the `get_total_balance` function from the spec:

https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#get_total_balance